bitswap/server/server.go

```diff
@@ -508,18 +501,16 @@ func (bs *Server) provideCollector(ctx context.Context) {
 	}
 }
 
-func (bs *Server) provideWorker(px process.Process) {
-	// FIXME: OnClosingContext returns a _custom_ context type.
-	// Unfortunately, deriving a new cancelable context from this custom
-	// type fires off a goroutine. To work around this, we create a single
-	// cancelable context up-front and derive all sub-contexts from that.
-	//
-	// See: https://github.com/ipfs/go-ipfs/issues/5810
-	ctx := procctx.OnClosingContext(px)
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-
+func (bs *Server) provideWorker(ctx context.Context) {
 	limit := make(chan struct{}, provideWorkerMax)
+	defer func() {
+		// Wait until all limitGoProvide goroutines are done before declaring
+		// this worker as done.
+		for i := 0; i < provideWorkerMax; i++ {
+			limit <- struct{}{}
+		}
+		bs.waitWorkers.Done()
+	}()
```

Comment:
I reviewed the provideWorker() closely and I think it works as before. I'm just not very sure why we have 1 provideWorker which then launches short-lived provideWorkerMax-goroutines all the time, vs having provideWorkerMax workers that are long-lived and read from the provideKeys channel and on closing we wait for them to finish with a waitGroup.

Not worth refactoring here though.

Reply:
The only reason this was done in the PR was to work exactly as before, with only the removal of goprocess.  Now that you have seen it too, and have the same question I did, I would like to change this to uses a set of long-lived goroutines.

Reply:
Do it in a different PR? This one fits its purpose already.

Reply:
I made the change in this PR.

Reply:
LGTM